documentation/writing_plugins.md

```diff
@@ -1,151 +1,492 @@
+- `resolve_reference`
+
+  This hook is used by the reference plugin itself. Most plugins will use this
+  hook.
+
+- `validate_resolve_reference`
+
+  This hook is used to pre-validate the parameters that will be passed to the
+  `resolve_reference` plugin. These plugins let Bolt raise any validation
+  errors during inventory loading, rather than in the middle of a plan run.
+  Regardless of whether this hook is specified, Bolt tests parameters to
+  make sure they match the parameters for the `resolve_reference` plugin.
+
+📖 **Related information**
+
+- [Example: yaml
+  plugin](https://github.com/puppetlabs/puppetlabs-yaml/tree/master/tasks)
+
+### Secret plugins
+
+Secret plugins are used to create keys for encryption and decryption, encrypt
+plaintext, and decrypt ciphertext. Bolt can be configured to use specific
+secret plugins when it runs the `bolt secret` command.
+
+There are three plugin hooks associated with secret plugins:
+
+- `secret_createkeys`
+
+  This hook is used for a plugin that creates keys for encryption and
+  decryption. Bolt always passes a `force` parameter to plugins using this hook.
+  When running the `bolt secret createkeys --force` command or
+  `New-BoltSecretKey -Force` cmdlet, the `force` parameter will have a value of
+  `true`.
+
+- `secret_decrypt`
+
+  This hook is used for a plugin that decrypts ciphertext and returns plaintext.
+  Bolt always passes an `encrypted_value` parameter to plugins using this hook,
+  which is set by running the `bolt secret decrypt` command or
+  `Unprotect-BoltSecret` cmdlet.
+
+- `secret_encrypt`
+
+  This hook is used for a plugin that encrypts plaintext and returns ciphertext.
+  Bolt always passes a `plaintext_value` parameter to plugins using this hook,
+  which is set by running the `bolt secret encrypt` command or
+  `Protect-BoltSecret` cmdlet.
+
+📖 **Related information**
+
+- [Example: pkcs7
+  plugin](https://github.com/puppetlabs/puppetlabs-pkcs7/tree/master/tasks)
 
-A plugin can implement hooks as tasks. Bolt will search for tasks in a plugin
-module matching the hook name, such as `my_plugin::resolve_reference`.
-Alternatively, you can map hooks to tasks in the `hooks` section of
-`bolt_plugin.json`. The `hooks` section is a JSON object where the keys are hook
-names and the values are objects with a key of `task` and a value containing the
-name of a task to run. You only have to specify a hook mapping in the `hooks`
-section if the task name does not match the hook name.
+### Puppet library plugins
+
+Bolt uses a Puppet library plugin to make sure the Puppet library is available
+on a target when the `apply_prep` function is called.
+
+Puppet library plugins behave differently from other types of plugins. While
+other types of plugins are run on `localhost`, Puppet library plugins are run on
+each target that you are running `apply_prep` on. Because of this, each target
+you run the plugin on must be able to interpret the scripting language the
+plugin uses.
```

Comment:
```suggestion
Puppet library plugins behave differently from other types of plugins. While
Bolt runs other types of plugins on `localhost`, Bolt runs Puppet library 
plugins on each target that you are running `apply_prep` on. 
Because of this behavior, each target you run the plugin on must be able 
to interpret the scripting language the plugin uses.
```